source/ptreferential/ptreferential_ng.cpp

```diff
@@ -0,0 +1,360 @@
+            >> ')';
+
+        get_corresponding = (qi::lit("get") | "GET") >> ident >> "<-" >> expr;
+        expr_leaf = '(' >> expr >> ')' | pred | get_corresponding;
+        expr_diff = (expr_leaf >> '-' >> expr_diff)
+            [_val = phx::construct<ast::BinaryOp<ast::Diff>>(_1, _2)]
+            | expr_leaf[_val = _1];
+        expr_and = (expr_diff >> (qi::lit("AND") | qi::lit("and")) >> expr_and)
+            [_val = phx::construct<ast::BinaryOp<ast::And>>(_1, _2)]
+            | expr_diff[_val = _1];
+        expr_or = (expr_and >> (qi::lit("OR") | qi::lit("or")) >> expr_or)
+            [_val = phx::construct<ast::BinaryOp<ast::Or>>(_1, _2)]
+            | expr_and[_val = _1];
+        expr = expr_or;
+
+        ident = qi::lexeme[qi::alpha >> *(qi::alnum | qi::char_("_"))];
+        str = qi::lexeme[+(qi::alnum | qi::char_("_.:;|-"))]
+            | qi::lexeme['"' > *(qi::char_ - qi::char_("\"\\") | ('\\' > qi::char_)) > '"'];
+    }
+
+    // Pred
+    qi::rule<Iterator, ast::Expr(), ascii::space_type> pred;
+    qi::rule<Iterator, ast::All(), ascii::space_type> all;
+    qi::rule<Iterator, ast::Empty(), ascii::space_type> empty;
+    qi::rule<Iterator, ast::Fun(), ascii::space_type> fun;
+    qi::rule<Iterator, ast::Fun(), ascii::space_type> cmp;
+    qi::rule<Iterator, ast::Fun(), ascii::space_type> dwithin;
+
+    // Expr
+    qi::rule<Iterator, ast::GetCorresponding(), ascii::space_type> get_corresponding;
+    qi::rule<Iterator, ast::Expr(), ascii::space_type> expr_leaf;
+    qi::rule<Iterator, ast::Expr(), ascii::space_type> expr_and;
+    qi::rule<Iterator, ast::Expr(), ascii::space_type> expr_or;
+    qi::rule<Iterator, ast::Expr(), ascii::space_type> expr_diff;
+    qi::rule<Iterator, ast::Expr(), ascii::space_type> expr;
+
+    // lexemes
+    qi::rule<Iterator, std::string(), ascii::space_type> ident;
+    qi::rule<Iterator, std::string(), ascii::space_type> str;
+};
+
+const char* to_string(OdtLevel_e level) {
+    switch (level) {
+    case type::OdtLevel_e::scheduled: return "scheduled";
+    case type::OdtLevel_e::with_stops: return "with_stops";
+    case type::OdtLevel_e::zonal: return "zonal";
+    default: return "all";
+    }
+}
+
+OdtLevel_e odt_level_from_string(const std::string& s) {
+    if (s == "scheduled") {
+        return OdtLevel_e::scheduled;
+    } else if (s == "with_stops") {
+        return OdtLevel_e::with_stops;
+    } else if (s == "zonal") {
+        return OdtLevel_e::zonal;
+    } else {
+        return OdtLevel_e::all;
+    }
+}
+
+boost::posix_time::ptime from_datetime(const std::string& s) {
+    if (s.empty()) {
+        throw parsing_error(parsing_error::partial_error, "empty datetime is illegal");
+    }
+    if (s.back() != 'Z') {
+        throw parsing_error(parsing_error::partial_error,
+                            "only UTC datetime are allowed (must finish with `Z`): " + s);
+    }
+    try {
+        return boost::posix_time::from_iso_string(s.substr(0, s.size() - 1));
```

Comment:
Why do you require UTC datetime with the `Z` suffix as an input, but treat it as local datetime when converted to a ptime ? 

Boost should handle the ending `Z` for you.

Reply:
Also, this could be moved to `datetime.cpp` ?

Reply:
boost::posix_time doesn't parse timezone. Doing UTC only is for simplicity. Doing it explicitly is because, in navitia, without timezone specification, that's local datetime.